### ISSUE TITLE
Errors in beforeSuite don't get correctly reported and prevent onComplete

### ISSUE THREAD

**Issue by jmnsf** [webdriverio/webdriverio#1460]
## The problem

Returning a rejected promise in `beforeSuite` results in an internal reporter error and the `onComplete` callback not being executed.

## Environment

* WebdriverIO version: 4.2.1
* Node.js version: 6.2.1
* wdio testrunner in sync mode
* `"wdio-mocha-framework": "^0.2.13"`

## Link to Selenium/WebdriverIO 

Not sure where to get these logs, happy to provide them if elucidated. Console logs (note missing `running onComplete`):

```bash
jmnsf:~/Code/temp/wdiotest$ node_modules/.bin/wdio

Starting Selenium 2.45.0 on port 4444
Filename: /Users/jmnsf/Code/temp/wdiotest/node_modules/selenium-launcher/lib/selenium-server-standalone-2.45.0.jar
/Users/jmnsf/Code/temp/wdiotest/node_modules/webdriverio/build/lib/utils/ReporterStats.js:340
            testStats.state = 'fail';
                            ^

TypeError: Cannot set property 'state' of undefined
    at ReporterStats.testFail (/Users/jmnsf/Code/temp/wdiotest/node_modules/webdriverio/build/lib/utils/ReporterStats.js:340:29)
    at BaseReporter.<anonymous> (/Users/jmnsf/Code/temp/wdiotest/node_modules/webdriverio/build/lib/utils/BaseReporter.js:146:25)
    at emitOne (events.js:96:13)
    at BaseReporter.emit (events.js:188:7)
    at BaseReporter.handleEvent (/Users/jmnsf/Code/temp/wdiotest/node_modules/webdriverio/build/lib/utils/BaseReporter.js:287:27)
    at Launcher.messageHandler (/Users/jmnsf/Code/temp/wdiotest/node_modules/webdriverio/build/lib/launcher.js:519:28)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at process.nextTick (internal/child_process.js:719:12)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
running after
```

## Code To Reproduce Issue

The config:

```javascript
let selenium;

const seleniumLauncher = require('selenium-launcher');

exports.config = {
    specs: [
        './test/**/*.js'
    ],
    exclude: [
    ],
    maxInstances: 1,
    capabilities: [{
        maxInstances: 1,
        browserName: 'phantomjs'
    }],
    sync: true,
    logLevel: 'silent',
    coloredLogs: true,
    screenshotPath: './errorShots/',
    baseUrl: 'http://localhost:8000',
    waitforTimeout: 10000,
    connectionRetryTimeout: 90000,
    connectionRetryCount: 3,
    framework: 'mocha',
    mochaOpts: {
        ui: 'bdd'
    },
    onPrepare: function (config, capabilities) {
        return new Promise(function (resolve, reject) {
            seleniumLauncher(function (err, s) {
                if (err) return reject(err);
                selenium = s;
                resolve();
            });
        });
    },
    beforeSuite: function (suite) {
        return Promise.reject(new Error('oh no'));
    },
    after: function (result, capabilities, specs) {
        console.log('running after');
    },
    onComplete: () => {
        console.log('running onComplete');
        if (selenium) selenium.kill();
    }
}
```

**Comment by christian-bromann** [comment]
@jmnsf why don't you use the [wdio-selenium-standalone-service](https://github.com/webdriverio/wdio-selenium-standalone-service) to start the selenium server before your tests begin? I added a fix that should prevent

**Comment by jmnsf** [comment]
@christian-bromann wasn't aware of it :)

Thanks for the fix and suggestion.

**Comment by christian-bromann** [comment]
My first fix broke our integration build at Sauce Labs. Will have to find a better solution + adequate tests.

**Comment by christian-bromann** [comment]
Your issue was triggered by an user hook like the ones you can define in your wdio config. These should never break the test process if they fail. Any service you hook into your test suite should be ignored in case of a failure (an error message will be prompt though). You returned a rejected promise which was not ignored and since the beforeSuite hook gets executed before any suite was registered that failure popped up in the reporter. I fixed the issue in the mocha adapter and will release that version soon.

**Comment by megagon** [comment]
still reproducibale @"webdriverio": "^4.5.1",
```/node_modules/webdriverio/build/lib/utils/ReporterStats.js:448
            testStats.state = 'fail';
```

**Comment by christian-bromann** [comment]
@megagon there will be a release today that will finally fix this

**Comment by megagon** [comment]
@christian-bromann thank you!

**Comment by christian-bromann** [comment]
Just publish `v4.5.2` with an ultimate fix for this. Let me know if you guys still encounter this.

**Comment by megagon** [comment]
looks like fixed, thank you!

**Comment by megagon** [comment]
have it a bit different now:
`node_modules/webdriverio/build/lib/utils/ReporterStats.js:442
            testStats.state = 'fail';
`

**Comment by christian-bromann** [comment]
@megagon did you update your spec reporter? Can you provide a reproducible example?

**Comment by megagon** [comment]
@christian-bromann spec reported was 0.0.4. I will update it and then we'll see, maybe it will be solved

**Comment by ignachiou** [comment]
@christian-bromann Hey men, i want to ask if you fix this? I still get it using v4.6.2
